src/core/experimental/array.d

```diff
@@ -0,0 +1,2559 @@
+        {
+            double newCapacity = capacity ? capacity * capacityFactor : stuff.length;
+            while (newCapacity < capacity + stuff.length)
+            {
+                newCapacity = newCapacity * capacityFactor;
+            }
+            reserve((() @trusted => cast(size_t)(newCapacity))());
+        }
+        //_support[pos + stuff.length .. _payload.length + stuff.length] =
+            //_support[pos .. _payload.length];
+        for (size_t i = _payload.length + stuff.length - 1; i >= pos +
+                stuff.length; --i)
+        {
+            // Avoids underflow if payload is empty
+            _support[i] = _support[i - stuff.length];
+        }
+
+        //writefln("typeof support[i] %s", typeof(_support[0]).stringof);
+
+        //_support[pos .. pos + stuff.length] = stuff[];
+        for (size_t i = pos, j = 0; i < pos + stuff.length; ++i, ++j) {
+            _support[i] = stuff[j];
+        }
+
+        _payload = (() @trusted => cast(T[])(_support[0 .. _payload.length + stuff.length]))();
+        return stuff.length;
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        Array!int a;
+        assert(a.empty);
+
+        size_t pos = 0;
+        pos += a.insert(pos, 1);
+        pos += a.insert(pos, [2, 3]);
+        assert(equal(a, [1, 2, 3]));
+    }
+
+    /**
+     * Check whether there are no more references to this array instance.
+     *
+     * Returns:
+     *      `true` if this is the only reference to this array instance;
+     *      `false` otherwise.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe
+    bool isUnique(this _)()
+    {
+        if (_support !is null)
+        {
+            return cast(bool) opCmpPrefix!"=="(_support, 1);
+        }
+        return true;
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        auto a = Array!int(24, 42);
+
+        assert(a.isUnique);
+        {
+            auto a2 = a;
+            assert(!a.isUnique);
+            a2.front = 0;
+            assert(a.front == 0);
+        } // a2 goes out of scope
+        assert(a.isUnique);
+    }
+
+    /**
+     * Check if the array is empty.
+     *
+     * Returns:
+     *      `true` if there are no elements in the array; `false` otherwise.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    @nogc nothrow pure @safe
+    bool empty() const
+    {
+        return length == 0;
+    }
+
+    ///
+    static if (is(T == int))
+    @safe unittest
+    {
+        Array!int a;
+        assert(a.empty);
+        size_t pos = 0;
+        a.insert(pos, 1);
+        assert(!a.empty);
+    }
+
+    /**
+     * Provide access to the first element in the array. The user must check
+     * that the array isn't `empty`, prior to calling this function.
+     *
+     * Returns:
+     *      a reference to the first element.
+     *
+     * Complexity: $(BIGOH 1).
+     */
+    ref auto front(this _)()
```

Comment:
Why the this template parameter?

Reply:
So all the `Array!T` types, regardless of the qualifiers can call front. It can't be `ref auto front() const` because mutable arrays must be able to change the element using front, like `a.front = 42;`

Reply:
So it’s just for attribute inference? In that case you can use an empty template parameter list instead. Perhaps add a comment as well.

Reply:
This is now gone.